sdk/nodejs/runtime/invoke.ts

```diff
@@ -251,28 +251,35 @@ export function call<T>(tok: string, props: Inputs, res?: Resource): Output<T> {
 }
 
 /**
- * Behaves exactly like `call` but returns a Promise instead. Unknowns are not allowed in the response, and an exception
- * will be thrown blaming the provider for violating the contract if the provider returns any unknown values. Secret
- * bits and dependencies returned from the provider are similarly discarded.
+ * Behaves exactly like `call` but returns a Promise instead, to support code generated calls for methods with
+ * XReturnPlainResource. After awaiting the output, extract the resource by field name specified in plainResourceField.
+ * Unknowns are not allowed in the response, and an exception will be thrown blaming the provider for violating the
+ * contract if the provider returns any unknown values. Secret bits and dependencies returned from the provider are
+ * similarly discarded at the moment.
  */
-export function callAsync<T>(tok: string, props: Inputs, res?: Resource): Promise<T> {
+export function callAsync<T>(tok: string, props: Inputs, res: Resource, callAsyncOpts: {
+    plainResourceField: string,
+}): Promise<T> {
     const label = `Calling function: tok=${tok} (callAsync)`;
     log.debug(label + (excessiveDebugOutput ? `, props=${JSON.stringify(props)}` : ``));
 
     return debuggablePromise(new Promise<T>((resolve, reject) => {
-        const resolver = (v: T, _isKnown: boolean, _isSecret: boolean, _deps?: Resource[], err?: Error | undefined) => {
+        const resolver = (v: T, isKnown: boolean, _isSecret: boolean, _deps?: Resource[], err?: Error | undefined) => {
             if (err) {
                 return reject(err);
             }
 
-            // For objects we recur into each field and transform it into an Output, which recursively flattens any
-            // nested Output. This is necessary for type safety, since the expected return types from Call methods
-            // expose properties as prompt values, but the runtime values such as v.foo here are Outputs.
-            //
-            // Unlike call, callAsync does not just transform v into an Output at top level to avoid unkown poisoning,
-            // the situation when any output field being unknown results in the entire result being unknown and
-            // corresponding loss of information. Instead, unknown values are discarded.
-            unsafeUnwrapProperties(v).then(resolve);
+            if (!isKnown) {
```

Comment:
Adding this for symmetry with Python and Go, but actually I'm wondering if this branch is dead.

There is a comment below stating that:

```
                // If the value the engine handed back is or contains an unknown value, the resolver will mark its value as
                // unknown automatically, so we just pass true for isKnown here. Note that unknown values will only be
                // present during previews (i.e. isDryRun() will be true).
```

So quite possibly this resolver is always called with `isKnown: true`.

Reply:
Ah okay, so this was wrong. I've edited this below.